lib/idle_task_queue.dart

```diff
@@ -0,0 +1,85 @@
+// @JS()
+library idle_task_queue;
+
+import 'dart:async';
+import 'dart:html';
+import 'dart:js';
+
+import 'package:meta/meta.dart';
+
+/// A handler which will be called during idle time.
+typedef FutureOr<R> IdleTaskHandler<R>();
+
+@immutable
+class _IdleTask<R> {
+  final IdleTaskHandler<R> handler;
+  final Completer<R> completer = new Completer<R>();
+
+  _IdleTask(this.handler);
+}
+
+/// Adds the task to a queue which will be called during idle time.
+///
+/// Returns the `Future` which will complete when the task has run.
+///
+/// May complete with an error if the task throws when called.
+Future<R> enqueueIdleTask<R>(IdleTaskHandler<R> handler) {
+  // For unsupported browsers when the polyfill is not available,
+  // we still want the task to be called eventually.
+  if (!_doesBrowserSupportIdleCallback) {
+    return new Future.delayed(Duration.ZERO).then((_) async {
```

Comment:
Following up the conversation here
> __Greg:__ #nit _immediately is a little misleading here, since this happens in the next tick. Also, the async modifier on this function also results in an additional async gap in Dart 1.
> 
> __Todd:__ In Dart 2 won't that cause the start of handler to be called synchronously?

No, the `new Future(callback)` constructor _"Creates a future containing the result of calling computation asynchronously with Timer.run."_ ([docs](https://api.dartlang.org/stable/2.1.0/dart-async/Future/Future.html)).

So, the function isn't run until, at the soonest, the next tick in both Dart 1 and Dart 2.